File: priv/combined_metrics/samples/scope_and_assignment/used_only_once/good/query_builder.py
"""Query/struct builder — GOOD: intermediate results are inlined."""

from urllib.parse import urlencode


class QueryBuilder:
    def build_search_query(self, filters):
        return (
            _from("products")
            .where(category=filters["category"])
            .where_lte("price", filters["max_price"])
            .where_gt("stock", 0)
            .order_by("inserted_at")
            .limit(filters["limit"])
        )

    def build_user_dict(self, attrs):
        return {
            "name": attrs["name"].strip(),
            "email": attrs["email"].lower(),
            "role": attrs.get("role") or "guest",
        }

    def format_report(self, data):
        header = f"=== {data['title'].upper()} ==="
        body = "\n".join(self._format_row(r) for r in data["rows"])
        return f"{header}\n{body}"

    def build_notification(self, event):
        return {
            "subject": f"Event: {event['name']}",
            "to": event["user"]["email"],
            "body": _render_template(_load_template(event["type"]), event),
        }

    def compose_url(self, base_url, path, query_params):
        return f"{base_url}{path}?{urlencode(query_params)}"

    def _format_row(self, row):
        return f"{row['label']}: {row['value']}"


def _load_template(t):
    return f"template_{t}"


def _render_template(t, _e):
    return t


class _from:
    def __init__(self, _t):
        pass

    def where(self, **_):
        return self

    def where_lte(self, *_):
        return self

    def where_gt(self, *_):
        return self

    def order_by(self, *_):
        return self

    def limit(self, *_):
        return self

File: priv/combined_metrics/samples/scope_and_assignment/used_only_once/good/query_builder.go
// Query/struct builder — GOOD: intermediate results are inlined or chained.
package querybuilder

import (
	"fmt"
	"net/url"
	"strings"
)

type Filters struct {
	Category string
	MaxPrice float64
	Limit    int
}

type User struct {
	Name  string
	Email string
	Role  string
}

type Row struct {
	Label string
	Value string
}

type ReportData struct {
	Title string
	Rows  []Row
}

type Event struct {
	Name string
	Type string
	User struct{ Email string }
}

type QueryBuilder struct{}

func (QueryBuilder) BuildSearchQuery(f Filters) *Query {
	return From("products").
		Where("category", f.Category).
		WhereLte("price", f.MaxPrice).
		WhereGt("stock", 0).
		OrderBy("inserted_at").
		Limit(f.Limit)
}

func (QueryBuilder) BuildUser(name, email, role string) User {
	if role == "" {
		role = "guest"
	}
	return User{
		Name:  strings.TrimSpace(name),
		Email: strings.ToLower(email),
		Role:  role,
	}
}

func (b QueryBuilder) FormatReport(d ReportData) string {
	header := fmt.Sprintf("=== %s ===", strings.ToUpper(d.Title))
	lines := make([]string, len(d.Rows))
	for i, r := range d.Rows {
		lines[i] = b.formatRow(r)
	}
	return header + "\n" + strings.Join(lines, "\n")
}

func (QueryBuilder) BuildNotification(e Event) map[string]string {
	return map[string]string{
		"subject": fmt.Sprintf("Event: %s", e.Name),
		"to":      e.User.Email,
		"body":    renderTemplate(loadTemplate(e.Type), e),
	}
}

func (QueryBuilder) ComposeURL(baseURL, path string, params url.Values) string {
	return fmt.Sprintf("%s%s?%s", baseURL, path, params.Encode())
}

func (QueryBuilder) formatRow(r Row) string {
	return fmt.Sprintf("%s: %s", r.Label, r.Value)
}

func loadTemplate(t string) string         { return "template_" + t }
func renderTemplate(t string, _ Event) string { return t }

type Query struct{}

func From(_ string) *Query                       { return &Query{} }
func (q *Query) Where(_, _ interface{}) *Query   { return q }
func (q *Query) WhereLte(_ string, _ float64) *Query { return q }
func (q *Query) WhereGt(_ string, _ int) *Query  { return q }
func (q *Query) OrderBy(_ string) *Query         { return q }
func (q *Query) Limit(_ int) *Query              { return q }
